core/src/main/java/jenkins/util/io/PathRemover.java

```diff
@@ -0,0 +1,286 @@
+                    }
+                }
+                if (waitBetweenRetries != 0 || gcAfterFailedRemove)
+                    sb.append(" between attempts");
+            }
+            if (interrupted.get())
+                sb.append(". The delete operation was interrupted before it completed successfully");
+            sb.append('.');
+            interrupted.set(false);
+            return sb.toString();
+        }
+    }
+    
+    private static Optional<IOException> tryRemoveFile(@Nonnull Path path) {
+        try {
+            removeOrMakeRemovableThenRemove(path);
+            return Optional.empty();
+        } catch (IOException e) {
+            return Optional.of(e);
+        }
+    }
+
+    private static List<IOException> tryRemoveRecursive(@Nonnull Path path) {
+        List<IOException> accumulatedErrors = Util.isSymlink(path) ? new ArrayList<>() :
+                tryRemoveDirectoryContents(path);
+        tryRemoveFile(path).ifPresent(accumulatedErrors::add);
+        return accumulatedErrors;
+    }
+
+    private static List<IOException> tryRemoveDirectoryContents(@Nonnull Path path) {
+        if (!Files.isDirectory(path)) return Collections.emptyList();
+        List<IOException> accumulatedErrors = new ArrayList<>();
+        try (DirectoryStream<Path> children = Files.newDirectoryStream(path)) {
+            for (Path child : children) {
+                accumulatedErrors.addAll(tryRemoveRecursive(child));
+            }
+        } catch (IOException e) {
+            accumulatedErrors.add(e);
+        }
+        return accumulatedErrors;
+    }
+
+    private static void removeOrMakeRemovableThenRemove(@Nonnull Path path) throws IOException {
+        try {
+            Files.deleteIfExists(path);
+        } catch (IOException e) {
+            makeRemovable(path);
+            try {
+                Files.deleteIfExists(path);
+            } catch (IOException e2) {
+                // see https://java.net/projects/hudson/lists/users/archive/2008-05/message/357
+                // I suspect other processes putting files in this directory
+                if (Files.isDirectory(path)) {
+                    List<String> entries;
+                    try (Stream<Path> children = Files.list(path)) {
+                        entries = children.map(Path::toString).collect(Collectors.toList());
+                    }
+                    throw new CompositeIOException("Unable to remove directory " + path + " with directory contents: " + entries, e, e2);
+                }
+                throw new CompositeIOException("Unable to remove file " + path, e, e2);
+            }
+        }
+    }
+
+    private static void makeRemovable(@Nonnull Path path) throws IOException {
+        if (!Files.isWritable(path)) {
+            makeWritable(path);
+        }
+        /*
+         on Unix both the file and the directory that contains it has to be writable
```

Comment:
What about Windows? Is it needed too? Although I guess It does't hurt

Reply:
That's a good question actually. I just copied that comment during refactoring. If this is different on Windows, I'll see about adding a test.